consensus/src/memory_pool.rs

```diff
@@ -18,244 +18,158 @@
+    serial_numbers: IndexSet<Digest>,
+    memos: IndexSet<Digest>,
 }
 
 const BLOCK_HEADER_SIZE: usize = BlockHeader::size();
 const COINBASE_TRANSACTION_SIZE: usize = 1490; // TODO Find the value for actual coinbase transaction size
 
-impl<T: TransactionScheme + Send + Sync + 'static> MemoryPool<T> {
+impl MemoryPool {
     /// Initialize a new memory pool with no transactions
     #[inline]
     pub fn new() -> Self {
         Self::default()
     }
 
-    /// Load the memory pool from previously stored state in storage
-    pub async fn from_storage<P: LoadableMerkleParameters, S: Storage>(
-        storage: &Ledger<T, P, S>,
-    ) -> Result<Self, ConsensusError> {
-        let memory_pool = Self::new();
-
-        if let Ok(Some(serialized_transactions)) = storage.get_memory_pool() {
-            if let Ok(transaction_bytes) = DPCTransactions::<T>::read(&serialized_transactions[..]) {
-                for transaction in transaction_bytes.0 {
-                    let size = transaction.size();
-                    let entry = Entry {
-                        transaction,
-                        size_in_bytes: size,
-                    };
-                    memory_pool.insert(storage, entry).await?;
-                }
-            }
-        }
-
-        Ok(memory_pool)
-    }
-
-    /// Store the memory pool state to the database
-    #[inline]
-    pub fn store<P: LoadableMerkleParameters, S: Storage>(
-        &self,
-        storage: &Ledger<T, P, S>,
-    ) -> Result<(), ConsensusError> {
-        let mut transactions = DPCTransactions::<T>::new();
-
-        for (_transaction_id, entry) in self.transactions.inner().iter() {
-            transactions.push(entry.transaction.clone())
-        }
-
-        let serialized_transactions = to_bytes![transactions]?.to_vec();
-
-        storage.store_to_memory_pool(serialized_transactions)?;
-
-        Ok(())
-    }
-
     /// Adds entry to memory pool if valid in the current ledger.
-    pub async fn insert<P: LoadableMerkleParameters, S: Storage>(
-        &self,
-        storage: &Ledger<T, P, S>,
-        entry: Entry<T>,
-    ) -> Result<Option<Vec<u8>>, ConsensusError> {
-        let transaction_serial_numbers = entry.transaction.old_serial_numbers();
-        let transaction_commitments = entry.transaction.new_commitments();
-        let transaction_memo = entry.transaction.memorandum();
-
-        if has_duplicates(transaction_serial_numbers)
-            || has_duplicates(transaction_commitments)
-            || self.contains(&entry)
+    pub fn insert(
+        &mut self,
+        ledger: &DynLedger,
+        transaction: SerialTransaction,
+    ) -> Result<Option<Digest>, ConsensusError> {
+        let transaction_id: Digest = transaction.id.into();
+
+        if has_duplicates(&transaction.old_serial_numbers)
+            || has_duplicates(transaction.new_records.iter().map(|x| &x.commitment))
+            || self.transactions.contains_key(&transaction_id)
```

Comment:
This check should probably be moved as the first one, as it's likely to be the fastest one, giving us nice short-circuiting here.

Reply:
it is first?

Reply:
I meant `self.transactions.contains_key(&transaction_id)` specifically, which could short-circuit the `x || y || z` predicate.